### DISCUSSION TITLE
Rich Text / PDF Render and Use as a Resume

### DISCUSSION THREAD

**Discussion by ctjlewis** [lowlighter/metrics#245] [Ideas]
Enough hard work has been put into making this tool that, looking at the output, I would be willing to add it as a supplicant to my resume or even submit in lieu of an actual resume.

Has anyone else considered this? What might this implementation look like, and what small details might be helpful to add to support this use case?

**Comment by lowlighter** [comment]
I guess the easiest way would be to create multiples jobs like in this [workflow](https://github.com/lowlighter/lowlighter/blob/master/.github/workflows/metrics.yml) using `config_output` with `png` or `jpg` to generate different images from plugins you're interested and embed them on an existing resume.

A more experimental and boldy way would be to start off from a [markdown template](https://github.com/lowlighter/lowlighter/blob/master/metrics.markdown.full.md), fill it up with text and template it with data gathered by metrics and rendered plugins images. With <kbd>Ctrl + P</kbd> on a browser you could get a PDF file from generated markdown. 
Actually, if we plugged metrics to LinkedIn API to get job data it may be possible to have a fully automated resume builder with day-to-day data 🙂

**Comment by lowlighter** [comment]
So I digged a bit about an eventual LinkedIn plugin, but it seems impossible to implement unfortunately 😕 (at least atm)

Basically there are no personal tokens so it's required to pass through an OAuth 2.0 token which is a pain to setup when you just want a token for yourself (+ it seems to be valid only for 2 months max so double pain). 

Even if you pass through an app, maximum permissions you can get as lambda user is `r_liteprofile` or `r_basicprofile` if you're lucky (and  you must create a "fictive company"), but to make something usable by metrics [`r_fullprofile`](https://docs.microsoft.com/en-us/linkedin/shared/references/v2/profile/full-profile?context=linkedin/consumer/context) would be required. 

It's clearly not designed for individual usage plus documentation seems a bit evasive and there's some broken links a bit everywhere on the developer portal 😞 I understand that they're not giving access easily to whole userbase but kind of bummer that you can't even get back your own data...

A solution based on puppeteer seems impossible too because of the authwall, and if using some kind of hack to log yourself you could just get banned so it's a no-go 😅 

Best that can be done is to setup an app which has to be approved to get `r_basicprofile` and get `headline` which is basically your current job name, but it's not worth the hassle.

I sent a support ticket to LinkedIn to see if there's any way to do it but I'm pretty sure it'll be a negative answer (if I get any)

**TLDR; no fully automated resume on GitHub readme profile with LinkedIn for now 🙂**

**Reply by ctjlewis** [reply]
Honestly Simon, I don't use LinkedIn, it kind of sucks. I'd sooner hand this in.

My suggestion was to maybe have a way to render this into a resume-like PDF, which I'd be happy to help with. Same information, just with a name and date at the top, contact info, etc.

Could always take a "work history" config option too to do most of the LinkedIn legwork without dealing too much with it.

If I wanted to provide this as a resume, I'd have to print the GitHub page out (with the full GH UI and so on), and it would not be a very clean approach. Mostly hitting that problem is what I'm thinking about, almost just a "print" option, to render to a text/PDF doc with all the same info/graphics + contact info.

**Reply by lowlighter** [reply]
Then I think the best would be to do something similar as stated in second part of https://github.com/lowlighter/metrics/discussions/245#discussioncomment-632604 and add a new util `svg.pdf` method based on the [`page.pdf`](https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#pagepdfoptions) from puppeteer and support `config_output: markdown-pdf`.

It shouldn't require too much heavy work to integrate this feature I think.

We'd do the render in headless browser like the current svg resizer/converter, but instead of returning raw svg (or using `page.screenshot` when converting to jpg/png) we call `page.pdf` to send back a pdf instead:
https://github.com/lowlighter/metrics/blob/553a1a1c3823201c20f74db237f018eb996e90f8/source/app/metrics/utils.mjs#L254-L258

To do so, the best would be to branch-out from the markdown pipeline just before the end of it and call this new `svg.pdf` method:
https://github.com/lowlighter/metrics/blob/553a1a1c3823201c20f74db237f018eb996e90f8/source/app/metrics/index.mjs#L90-L91

Since we have internal markdown interpretation since #237 and since GitHub style is opensourced (primer style) we could just load it using `page.addStyleTag` (or similar) before saving to pdf.

Finally `metadata.yml` of core plugin needs to be patched to allow `markdown-pdf` in allowed values:
https://github.com/lowlighter/metrics/blob/553a1a1c3823201c20f74db237f018eb996e90f8/source/plugins/core/metadata.yml#L184-L187

Of course at least one test and some documentation update (both in plugins/core) would be nice 🙂 

Also, this solution has the merit to work both in action and web instances.

To test in dev, one would need to setup a local web instance and appending something like below:
`?config.output=markdown-pdf&template=markdown&markdown=urlEncodedLinkToRawGitHubUserContentOfTemplate`
(all options are almost same as action, except it's dot.case instead of snake_case)
___

So basically users would need to create a markdown template where they can add their presentation, custom text and formatting, contact info, metrics plugins and other stuff like they want.

Then after configuring metrics, template would be rendered and outputed as pdf and pushed (action) or available as download (web). 

___

Let me know if you think this is a suitable solution 🙂

**Reply by lowlighter** [reply]
I'll start working on it later today, unless if you want to make the PR yourself 🙂

**Reply by ctjlewis** [reply]
Simon, so sorry, I have been insanely overworked the last few days! I think your implementation is perfect, and I feel bad not contributing here, but I'm absolutely slammed for time. I'm super thankful for your work and sorry I didn't respond soon.

I could help with small touches or styling on this, and will more likely *follow-up* on your work here. In the meantime, I might be more help with a lighter feature than this one.

Specifically, once we can render this to PDF, the final touch is being able to select Featured Issues and PRs, which will just render as a grid of GitHub previews. I'll even feature this issue where we worked out how to do the PDF render! Once that's available, there's nothing I could want on the document that I couldn't put there, and render to a resume.

You've done an outstanding job here and the configuration documentation is the best I've ever seen.

**Reply by lowlighter** [reply]
That's fine no need to be sorry, it's still a hobby project so it's not as if there was any deadlines or forced contributions 😄 

I actually "drafted" this pdf feature yesterday (in #270 and 16d455c) and it looks like [this](https://github.com/lowlighter/lowlighter/blob/master/metrics.markdown.pdf) which renders from [this markdown template](https://github.com/lowlighter/metrics/blob/master/source/templates/markdown/example.pdf.md) *(Note that raw version has working links and render seems better than the one displayed by GitHub document previewer)*

But I'm not sure how stable it is yet 🤔 
For example I needed to use `config_twemoji: yes` because emojis didn't render well (probably a font issue) but it makes me wondering how well embed SVGs renders across all platforms.

There's probably more tests and experimentations to do in order to detect some bugs.
I'm not a PDF expert (I just use them like most of people to share documents) so not sure if there's additional features that could be added without too much trouble like ensuring that non-latin characters works as expected, lock down the pdf to prevent editions, metadata/headers filling, etc.

___

> I could help with small touches or styling on this, and will more likely follow-up on your work here. In the meantime, I might be more help with a lighter feature than this one.

Yes any help is appreciated and there's still a lot of room for improvements 🙂 

A featured issues and prs would be nice indeed, and maybe another plugin to display pinned/selected repositories would be useful too if you intend to show featured works. There's some help on how you can extend metrics with new plugins in [contributing.md](https://github.com/lowlighter/metrics/blob/master/CONTRIBUTING.md) hopefully it's clear enough (if not you can always ask help in discussions)

Thanks a lot for your support 😊 !

**Reply by ctjlewis** [reply]
No problem at all! You're doing really good work.

I tried to set the output to `markdown-pdf` but it lags for a long time on render and eventually fails with `fonts is not defined`. I might have configured it incorrectly? https://github.com/ctjlewis/ctjlewis/runs/2447723705?check_suite_focus=true

**Reply by lowlighter** [reply]
Yes docs are not explicit yet 😅 

You need at least the following options for now (probably in future it could be simplified a bit):
```yml
with:
  token: ${{ secrets.METRICS_TOKEN }}
  config_output: markdown-pdf 
  # Currently only markdown derived templates are supported
  template: markdown  
  # File extensions still default to svg I think so it's better if you explicitely set it to pdf
  filename: metrics.markdown.pdf 
  # Url to your markdown template
  markdown: https://raw.githubusercontent.com/lowlighter/metrics/master/source/templates/markdown/example.pdf.md    
  # (optional) Folder where embed outputs are stored 
  markdown_cache: .cache 
  # (optional) If you want emojis
  config_twemoji: yes 
```

Markdown template works a bit differently than the SVG ones, basically you write your text as you'd do in a markdown file, and when you want to embed a plugin you include it with `<%- await embed("unique-filename", {options}) %>`

For example for stackoverflow the following should work:
```ejs
<%- await embed("stackoverflow", {
  stackoverflow:true, 
  stackoverflow_user:1674992, 
  stackoverflow_limit:3, 
  stackoverflow_lines:4
}) %>
```

**Reply by ctjlewis** [reply]
Ahhhh it works! I will work on the template!

Does the README give instructions on how to do a local run? I'd love to just throw my config in and run it locally, tweak the template as I go. I thought `yarn start` would cover this but not quite.

Will send PR.

---
Footnote: I tried to use [nektos/act](https://github.com/nektos/act) to debug the workflow locally but it complains about the `using: composite` flag.

**Reply by ctjlewis** [reply]
Also, so sorry, but I can't find the `example-isocalendar-pdf` plugin. Could you help me out? 😅 

Also, how would I go about adding the `base` section? I tried everything from:

```js
embed('base', { base: ... })
embed('isocalendar', { isocalendar: true })

// to:
embed('resume', { base: ..., isocalendar: true })
```

And have only managed to get the calendar to render. At first I thought `example-isocalendar-pdf` was its own plugin but that doesn't seem like the case now. 

Sorry for all the questions! I will add all this to the docs for you at some point based off whatever notes.

**Reply by lowlighter** [reply]
> Also, so sorry, but I can't find the example-isocalendar-pdf plugin. Could you help me out? 😅

The first argument is a unique string which gets used as filename when the action is run on GitHub (it gets stored in `markdown_cache` folder). AFAIK it's not possible to upload programmatically to GitHub camo and since base64 sources images don't render in markdown, metrics store images in target repository instead. So as you figured out it doesn't actually matters what you put in there 🙂 

> Also, how would I go about adding the base section?

I improved the inputs normalization in `embed` function, when you tested it only supported `dot.case` (web syntax). Now you should be able to use all syntaxes for `base`:
- `{base:"header, community, ..."}` (action syntax)
- `{"base.header":true, "base.community":true, ...}` (web syntax)
- `{base:["header", "community", ...]}` (parsed syntax)

Others option should works both in `snake_case` (action syntax) and `dot.case` without `plugin` prefixes (web syntax) 👍 
___

> Does the README give instructions on how to do a local run? I'd love to just throw my config in and run it locally, tweak the template as I go

Running locally in "action mode" has not been implemented properly 🙂  

Currently the `start` script make it run in web server mode because historically metrics was a web service first and later switched to an hybrid app. There is still a "hacky way" to do it which is used in [tests/metrics.test.js](https://github.com/lowlighter/metrics/blob/master/tests/metrics.test.js): 
https://github.com/lowlighter/metrics/blob/5b63ecbd96d79b4f81afadbc767f3a542019d24f/tests/metrics.test.js#L11-L27

Basically it parses `action.yml` (for defaults values), convert inputs to environment variables (e.g. `INPUT_*`) and then spawn a child process which acts the same as a "GitHub runner". 

Tests are only run in `dryrun` so there is actually no output. On GitHub runners, files get written to `/renders` which is mounted to  `/metrics_renders` folder. We could edit this so when run locally a CLI option lets it write somewhere else 🤔 ?
https://github.com/lowlighter/metrics/blob/5b63ecbd96d79b4f81afadbc767f3a542019d24f/source/app/action/index.mjs#L280-L286

I suppose this could be used as a workaround for now since composite actions are not supported yet by [nektos/act](https://github.com/nektos/act) as per https://github.com/nektos/act/issues/607. A PR has been opened for it but not sure when it's going to be merged

> Sorry for all the questions! I will add all this to the docs for you at some point based off whatever notes.

No problems! Hope it helped 🙂

**Reply by ctjlewis** [reply]
It definitely helped! I'm making progress, and got the PDF to render the sections I want.

Final thing: Would there be a way to inherit the top-level settings? Can this template read from the same variables in the workflow file that the profile render does? I think if it could inherit the defaults and allow for overrides with a `pdf_` prefix, i.e. `pdf_base: ['header', 'activity', ...]` (my profile by default does not include `header`), this would be pretty much close and shut on features and would be good to document and roll out. 

Once you can configure what goes on the PDF with arbitrary precision, and the template inherits the workflow settings by default, this is pretty much implemented. Also really, your work here has been absolutely first-class.

**Reply by lowlighter** [reply]
> Would there be a way to inherit the top-level settings?

I edited a bit the `embed` function so it inherits all top-level options by default (it makes more sense anyway) except "enable switches", so everything will be pre-configured but you can still choose when to display each plugin in template.

```yaml
with:
  base: header, repositories
  plugin_isocalendar_duration: full-year
```
```ejs
<!-- Isocalendar plugin, defaulting to "full-year" as per top-level definition -->
<%- await embed(`isocalendar`, {isocalendar:true}) %>
<!-- Isocalendar plugin, overriden to "half-year" -->
<%- await embed(`isocalendar-half`, {isocalendar:true, isocalendar_duration:"half-year"}) %>

<!-- Set base to "true" to default to top-level definition -->
<%- await embed(`base`, {base:true, notable_repositories:false}) %>
<!-- Or override base definition -->
<%- await embed(`base-custom`, {base:"community, activity"}) %>
```
I don't think `pdf_` options are needed with this inheritance system, since it's probably easier to edit the template directly

> Also really, your work here has been absolutely first-class.

Thanks a lot for the compliment 🙂

**Comment by ctjlewis** [comment]
I got this mostly working and it's [pretty beautiful](https://github.com/ctjlewis/ctjlewis/blob/master/resume-template.markdown.pdf) (GitHub PDF preview kinda messes it up but it looks fine on download). With a little HTML in the markdown, we can force page breaks, which is how this is blocked out (otherwise it just runs out into the next page):

```html
<div style="break-after: always; page-break-after: always;"></div>
<div style="padding-top: 3vh;"></div>
```

One issue I think I found: I can't override `activity_days`, I tried 0 and 365 days and can't get the window to change from 14 days, sadly. In this preview you'll notice the activity page renders the last two weeks of events, but `activity_days: 365` is set in the embed.

But it looks great honestly. The final thought I'm having is, can we get these to render as actual anchor links? The markdown links seem to work out of the box (which is amazing) but the GitHub metadata (like "repositories", links in activities sections) do not.

Also, finally, is there a way for me to see these `console.debug` statements and so on, even if I can't run it locally? Super sorry for the slew of questions again lol.

**Reply by lowlighter** [reply]
It looks pretty nice 🙂 !

> With a little HTML in the markdown, we can force page breaks

Yes it's a pretty nice trick to be able to make page-break, it's worth mentioning in a doc 👍 

> One issue I think I found: I can't override activity_days, I tried 0 and 365 days and can't get the window to change from 14 days, sadly.

It could be a bug indeed 😕 
Activity plugin currently only fetch first 100 events so it could be also be possible that none of them is older than 14 days.
We could switch to a loop instead like in coding habits plugin:
https://github.com/lowlighter/metrics/blob/b2871d99467aa53f697cd8ffb028df93b7e92ea9/source/plugins/habits/index.mjs#L21-L24

I don't how far you can get with GitHub API, but 365 days is pretty ambitious (though it may be possible) 🙂 

>  can we get these to render as actual anchor links

Templates doesn't have links because when embed through `<img>` svgs links doesn't work unfortunately.
I'm not sure whether we should add them in it since now there are [markdown plugins](https://github.com/lowlighter/lowlighter/blob/master/metrics.markdown.full.md). I think it makes more sense for heavy-text based stuff to use them instead, but there's less styling on it though

> Also, finally, is there a way for me to see these console.debug statements and so on, even if I can't run it locally?

Yes you can set `debug: yes`. Then in rendering section of action logs you can expand each rendering "Attempts" to see them 🙂

**Reply by ctjlewis** [reply]
Incredible. Thank you so much. A few days ago I started getting "an error occurred" for activity so I think the window is too wide.

**Reply by lowlighter** [reply]
If you set `debug: yes` on your workflow I can take a look at what causes the error if needed

**Reply by ctjlewis** [reply]
I merged from upstream and now the activities page is blank when I comment out the overrides, and the +- lines changed also errored:

https://github.com/ctjlewis/ctjlewis/blob/master/resume-template.markdown.pdf
https://github.com/ctjlewis/ctjlewis/runs/2558007075?check_suite_focus=true

**Reply by lowlighter** [reply]
I've pushed 421a770 to fix activity plugin:
https://raw.githubusercontent.com/lowlighter/ctjlewis/master/resume-template.markdown.pdf

As for lines and empty stats in community sections I'll check later (I forked your repositories for easier debug) 👍

**Reply by ctjlewis** [reply]
Thanks a bunch Simon. No rush on any of this, the activities section def used to work, but it seemed frozen to the default lookback. I tried to tinker around and couldn't get it unstuck sadly.

**Comment by lowlighter** [comment]
The upcoming `config_display` for v3.10 may be quite convenient for this feature as it'll allow to take advantage of the larger width of a pdf page and will feel more natural 🙂 

<img src="https://user-images.githubusercontent.com/22963968/118713145-e3c98b00-b821-11eb-8744-db98cfdd291f.png" width="700">

**Comment by jayvdb** [comment]
StackOverflow used to have an "Import from Linkedin", but it was removed because of Linkedin TOS. c.f. https://meta.stackoverflow.com/questions/312851/feature-request-import-cv-from-linkedin

While not as "good" as Linkedin, StackOverflow has a "developer story" feature which could be used as a data source.  I couldnt see if this has an API (it doesnt appear in https://api.stackexchange.com/docs) , but the HTML source is quite basic to parse (see source of https://stackoverflow.com/cv/jayvdb for example).  Ah, they have hresume support .. ?  c.f. https://meta.stackexchange.com/questions/28375/should-careers-stackoverflow-com-support-hresume/32216

**Reply by lowlighter** [reply]
This looks interesting, may be worth to dig deeper 👍
